File: test/test_did_dns.cpp
#include <doctest/doctest.h>

#include <string>
#include <vector>

#include "../include/authbox.hpp"

using namespace authbox::did;

TEST_CASE("did:dns - parse valid DID") {
    const std::string did_uri = "did:dns:example.com";

    auto domain = parse_did_dns(did_uri);
    REQUIRE(domain.is_ok());
    CHECK(domain.value() == "example.com");
}

TEST_CASE("did:dns - parse subdomain") {
    const std::string did_uri = "did:dns:identity.example.com";

    auto domain = parse_did_dns(did_uri);
    REQUIRE(domain.is_ok());
    CHECK(domain.value() == "identity.example.com");
}

TEST_CASE("did:dns - reject invalid method") {
    const std::string did_uri = "did:web:example.com";

    auto domain = parse_did_dns(did_uri);
    CHECK(domain.is_err());
}

TEST_CASE("did:dns - reject empty domain") {
    const std::string did_uri = "did:dns:";

    auto domain = parse_did_dns(did_uri);
    CHECK(domain.is_err());
}

TEST_CASE("did:dns - reject domain with spaces") {
    const std::string did_uri = "did:dns:example .com";

    auto domain = parse_did_dns(did_uri);
    CHECK(domain.is_err());
}

TEST_CASE("did:dns - reject domain without TLD") {
    const std::string did_uri = "did:dns:localhost";

    auto domain = parse_did_dns(did_uri);
    CHECK(domain.is_err());
}

TEST_CASE("did:dns - get DNS query domain") {
    const std::string domain = "example.com";
    const std::string query_domain = get_dns_query_domain(domain);

    CHECK(query_domain == "_did.example.com");
}

TEST_CASE("did:dns - resolve with mocked DNS returning JSON document") {
    const std::string did_uri = "did:dns:example.com";

    // Mock DID document
    const std::string mock_document = R"({
        "id": "did:dns:example.com",
        "@context": "https://www.w3.org/ns/did/v1",
        "verificationMethod": [{
            "id": "did:dns:example.com#key-1",
            "type": "JsonWebKey2020",
            "controller": "did:dns:example.com",
            "publicKeyJwk": {
                "kty": "OKP",
                "crv": "Ed25519",
                "x": "11qYAYKxCrfVS_7TyWQHOg7hcvPapiMlrwIaaPcHURo"
            }
        }],
        "authentication": ["did:dns:example.com#key-1"]
    })";

    // Create mock DNS lookup function
    DnsLookupFn mock_dns = [&](std::string_view domain) -> DidResult<std::vector<DnsTxtRecord>> {
        CHECK(domain == "_did.example.com");

        std::vector<DnsTxtRecord> records;
        DnsTxtRecord record;
        record.value = mock_document;
        record.dnssec_valid = true;
        records.push_back(record);

        return DidResult<std::vector<DnsTxtRecord>>::ok(std::move(records));
    };

    // Set up options with mock DNS
    DnsResolveOptions options;
    options.dns_lookup = mock_dns;

    // Resolve the DID
    auto doc_result = resolve_did_dns_document_json(did_uri, options);
    REQUIRE(doc_result.is_ok());
    CHECK(doc_result.value() == mock_document);
}

TEST_CASE("did:dns - resolve with no DNS lookup callback fails") {
    const std::string did_uri = "did:dns:example.com";

    DnsResolveOptions options;
    // No dns_lookup callback set

    auto doc_result = resolve_did_dns_document_json(did_uri, options);
    CHECK(doc_result.is_err());
}

TEST_CASE("did:dns - resolve with empty TXT records fails") {
    const std::string did_uri = "did:dns:example.com";

    // Mock DNS that returns no records
    DnsLookupFn mock_dns = [](std::string_view) -> DidResult<std::vector<DnsTxtRecord>> {
        return DidResult<std::vector<DnsTxtRecord>>::ok(std::vector<DnsTxtRecord>{});
    };

    DnsResolveOptions options;
    options.dns_lookup = mock_dns;

    auto doc_result = resolve_did_dns_document_json(did_uri, options);
    CHECK(doc_result.is_err());
}

TEST_CASE("did:dns - resolve with DNSSEC requirement") {
    const std::string did_uri = "did:dns:example.com";

    const std::string mock_document = R"({
        "id": "did:dns:example.com",
        "@context": "https://www.w3.org/ns/did/v1",
        "verificationMethod": [{
            "id": "did:dns:example.com#key-1",
            "type": "JsonWebKey2020",
            "controller": "did:dns:example.com",
            "publicKeyJwk": {
                "kty": "OKP",
                "crv": "Ed25519",
                "x": "11qYAYKxCrfVS_7TyWQHOg7hcvPapiMlrwIaaPcHURo"
            }
        }],
        "authentication": ["did:dns:example.com#key-1"]
    })";

    // Mock DNS with DNSSEC valid record
    DnsLookupFn mock_dns = [&](std::string_view) -> DidResult<std::vector<DnsTxtRecord>> {
        std::vector<DnsTxtRecord> records;
        DnsTxtRecord record;
        record.value = mock_document;
        record.dnssec_valid = true;
        records.push_back(record);
        return DidResult<std::vector<DnsTxtRecord>>::ok(std::move(records));
    };

    DnsResolveOptions options;
    options.dns_lookup = mock_dns;
    options.require_dnssec = true;

    auto doc_result = resolve_did_dns_document_json(did_uri, options);
    REQUIRE(doc_result.is_ok());
}

TEST_CASE("did:dns - resolve fails when DNSSEC required but not valid") {
    const std::string did_uri = "did:dns:example.com";

    const std::string mock_document = R"({"id": "did:dns:example.com"})";

    // Mock DNS with DNSSEC invalid record
    DnsLookupFn mock_dns = [&](std::string_view) -> DidResult<std::vector<DnsTxtRecord>> {
        std::vector<DnsTxtRecord> records;
        DnsTxtRecord record;
        record.value = mock_document;
        record.dnssec_valid = false;  // Not validated
        records.push_back(record);
        return DidResult<std::vector<DnsTxtRecord>>::ok(std::move(records));
    };

    DnsResolveOptions options;
    options.dns_lookup = mock_dns;
    options.require_dnssec = true;  // Require DNSSEC

    auto doc_result = resolve_did_dns_document_json(did_uri, options);
    CHECK(doc_result.is_err());
}

TEST_CASE("did:dns - integration with resolver") {
    const std::string did_uri = "did:dns:example.com";

    const std::string mock_document = R"({
        "id": "did:dns:example.com",
        "@context": "https://www.w3.org/ns/did/v1",
        "verificationMethod": [{
            "id": "did:dns:example.com#key-1",
            "type": "JsonWebKey2020",
            "controller": "did:dns:example.com",
            "publicKeyJwk": {
                "kty": "OKP",
                "crv": "Ed25519",
                "x": "11qYAYKxCrfVS_7TyWQHOg7hcvPapiMlrwIaaPcHURo"
            }
        }],
        "authentication": ["did:dns:example.com#key-1"]
    })";

    // Mock DNS lookup
    DnsLookupFn mock_dns = [&](std::string_view) -> DidResult<std::vector<DnsTxtRecord>> {
        std::vector<DnsTxtRecord> records;
        DnsTxtRecord record;
        record.value = mock_document;
        record.dnssec_valid = true;
        records.push_back(record);
        return DidResult<std::vector<DnsTxtRecord>>::ok(std::move(records));
    };

    // Set up options
    DnsResolveOptions dns_options;
    dns_options.dns_lookup = mock_dns;

    // Create resolver and register did:dns handler
    Resolver resolver;
    resolver.registry().register_method("dns", create_dns_handler(dns_options));

    // Resolve the DID
    auto resolution = resolver.resolve(did_uri);
    REQUIRE(resolution.is_ok());

    CHECK(std::string_view(resolution.value().document.id.data(), resolution.value().document.id.size()) ==
          "did:dns:example.com");
    CHECK(resolution.value().document.verification_methods.size() == 1);
}

TEST_CASE("did:dns - reject non-JSON TXT records") {
    const std::string did_uri = "did:dns:example.com";

    // Mock DNS returning invalid content
    DnsLookupFn mock_dns = [](std::string_view) -> DidResult<std::vector<DnsTxtRecord>> {
        std::vector<DnsTxtRecord> records;
        DnsTxtRecord record;
        record.value = "this is not json";
        record.dnssec_valid = true;
        records.push_back(record);
        return DidResult<std::vector<DnsTxtRecord>>::ok(std::move(records));
    };

    DnsResolveOptions options;
    options.dns_lookup = mock_dns;

    auto doc_result = resolve_did_dns_document_json(did_uri, options);
    CHECK(doc_result.is_err());
}

File: src/did/tests/test_did_dns.rs
use super::super::*;

const DID_DNS: &str = "did:dns:example.com";

fn mock_document() -> String {
    r#"{
        "id": "did:dns:example.com",
        "@context": "https://www.w3.org/ns/did/v1",
        "verificationMethod": [{
            "id": "did:dns:example.com#key-1",
            "type": "JsonWebKey2020",
            "controller": "did:dns:example.com",
            "publicKeyJwk": {
                "kty": "OKP",
                "crv": "Ed25519",
                "x": "11qYAYKxCrfVS_7TyWQHOg7hcvPapiMlrwIaaPcHURo"
            }
        }],
        "authentication": ["did:dns:example.com#key-1"]
    }"#
    .to_string()
}

#[test]
fn did_dns_parses_valid_domains_and_subdomains() {
    assert_eq!(parse_did_dns(DID_DNS).unwrap(), "example.com");
    assert_eq!(
        parse_did_dns("did:dns:identity.example.com").unwrap(),
        "identity.example.com"
    );
}

#[test]
fn did_dns_rejects_invalid_method_empty_domain_spaces_and_missing_tld() {
    assert!(parse_did_dns("did:web:example.com").is_err());
    assert!(parse_did_dns("did:dns:").is_err());
    assert!(parse_did_dns("did:dns:example .com").is_err());
    assert!(parse_did_dns("did:dns:localhost").is_err());
}

#[test]
fn did_dns_gets_txt_query_domain() {
    assert_eq!(get_dns_query_domain("example.com"), "_did.example.com");
}

#[test]
fn did_dns_txt_record_default_matches_cpp_aggregate_defaults() {
    let record = DnsTxtRecord::default();

    assert_eq!(record.value, "");
    assert!(!record.dnssec_valid);
}

#[test]
fn did_dns_resolves_with_mocked_dns_returning_json_document() {
    let document = mock_document();
    let options = DnsResolveOptions::with_lookup(move |domain| {
        assert_eq!(domain, "_did.example.com");
        Ok(vec![DnsTxtRecord {
            value: document.clone(),
            dnssec_valid: true,
        }])
    });

    let resolved = resolve_did_dns_document_json(DID_DNS, &options).unwrap();
    assert_eq!(resolved, mock_document());
}

#[test]
fn did_dns_skips_non_json_txt_and_resolves_later_dnssec_record() {
    let document = mock_document();
    let options = DnsResolveOptions::with_lookup(move |domain| {
        assert_eq!(domain, "_did.example.com");
        Ok(vec![
            DnsTxtRecord {
                value: "this is not json".to_string(),
                dnssec_valid: true,
            },
            DnsTxtRecord {
                value: document.clone(),
                dnssec_valid: true,
            },
        ])
    })
    .require_dnssec(true);

    let resolved = resolve_did_dns_document_json(DID_DNS, &options).unwrap();
    assert_eq!(parse_document(&resolved).unwrap().id, DID_DNS);
}

#[test]
fn did_dns_accepts_any_valid_json_txt_record_like_cpp_surface() {
    let json_record = r#"{"hello":"world"}"#;
    let options = DnsResolveOptions::with_lookup(move |_| {
        Ok(vec![DnsTxtRecord {
            value: json_record.to_string(),
            dnssec_valid: true,
        }])
    });

    let resolved = resolve_did_dns_document_json(DID_DNS, &options).unwrap();

    assert_eq!(resolved, json_record);
}

#[test]
fn did_dns_returns_original_json_txt_record_without_trimming() {
    let json_record = " \n {\"hello\":\"world\"}\t ";
    let options = DnsResolveOptions::with_lookup(move |_| {
        Ok(vec![DnsTxtRecord {
            value: json_record.to_string(),
            dnssec_valid: true,
        }])
    });

    let resolved = resolve_did_dns_document_json(DID_DNS, &options).unwrap();

    assert_eq!(resolved, json_record);
}

#[test]
fn did_dns_resolve_failures_match_cpp_surface() {
    assert!(resolve_did_dns_document_json(DID_DNS, &DnsResolveOptions::default()).is_err());

    let empty_records = DnsResolveOptions::with_lookup(|_| Ok(vec![]));
    assert!(resolve_did_dns_document_json(DID_DNS, &empty_records).is_err());

    let invalid_json = DnsResolveOptions::with_lookup(|_| {
        Ok(vec![DnsTxtRecord {
            value: "this is not json".to_string(),
            dnssec_valid: true,
        }])
    });
    assert!(resolve_did_dns_document_json(DID_DNS, &invalid_json).is_err());
}

#[test]
fn did_dns_respects_dnssec_requirement() {
    let valid = DnsResolveOptions::with_lookup(|_| {
        Ok(vec![DnsTxtRecord {
            value: mock_document(),
            dnssec_valid: true,
        }])
    })
    .require_dnssec(true);
    assert!(resolve_did_dns_document_json(DID_DNS, &valid).is_ok());

    let invalid = DnsResolveOptions::with_lookup(|_| {
        Ok(vec![DnsTxtRecord {
            value: mock_document(),
            dnssec_valid: false,
        }])
    })
    .require_dnssec(true);
    assert!(resolve_did_dns_document_json(DID_DNS, &invalid).is_err());
}

#[test]
fn did_dns_integrates_with_resolver() {
    let document = mock_document();
    let dns_options = DnsResolveOptions::with_lookup(move |_| {
        Ok(vec![DnsTxtRecord {
            value: document.clone(),
            dnssec_valid: true,
        }])
    });
    let mut resolver = Resolver::new();
    resolver
        .registry_mut()
        .register_method("dns", create_dns_handler(dns_options));

    let resolution = resolver.resolve(DID_DNS).unwrap();

    assert_eq!(resolution.document.id, DID_DNS);
    assert_eq!(resolution.document.verification_methods.len(), 1);
    assert_eq!(resolution.source_url, "dns:_did.example.com");
}

#[test]
fn did_dns_resolver_uses_configured_lookup_and_errors_without_one() {
    let document = mock_document();
    let resolver = Resolver::with_dns_lookup(move |domain| {
        assert_eq!(domain, "_did.example.com");
        Ok(vec![DnsTxtRecord {
            value: document.clone(),
            dnssec_valid: true,
        }])
    });

    let resolution = resolver.resolve(DID_DNS).unwrap();

    assert_eq!(resolution.document.id, DID_DNS);
    assert_eq!(resolution.document.verification_methods.len(), 1);
    assert_eq!(resolution.source_url, "dns:_did.example.com");

    let resolver_without_dns = Resolver::new();
    assert!(resolver_without_dns.resolve(DID_DNS).is_err());
}
